#' Simulate statistical power for EWAS.
#'
#' @importFrom grDevices dev.off pdf
#' @description
#' This function estimates the statistical power for each delta-beta of
#' methylation changes under given conditions using the `pwrEWAS` R package 
#' to identify an appropriate delta-beta cutoff. If the `plot` argument is 
#' set to `TRUE`, a `ggplot2` object representing the calculated statistical
#' power will also be returned.
#' 
#' Considering the statistical power and appropriate delta-beta cutoff is 
#' important to minimize false positive results. Even with adequately large 
#' sample sizes, we recommend considering probes with a methylation 
#' difference of 2% or more (|delta-beta| > 0.02) as statistically 
#' significant due to the technical variability of the EPIC array.
#'
#' Currently, the `pwrEWAS` R package is not available on Bioconductor, so 
#' users interested in using this function will need to install the 
#' `pwrEWAS` R package directly from GitHub : https://github.com/stefangraw/pwrEWAS
#'
#' 
#'
#' @param DMP.object A returned object from `MeCall.DMP()` or a data frame with "adjusted.p" and "deltabeta" as column names and CpG id as row names. Please note that the results should include information for all probes (set to `cutoff.P = 1`).
#' @param tissue.type A tissue type of sample ("Adult (PBMC)", "Saliva", "Sperm", "Lymphoma","Placenta", "Liver", "Colon", "Blood adult", "Blood 5 year olds", "Blood newborns", "Cord-blood (whole blood)", and "Cord-blood (PBMC)"). See manual of pwrEWAS R package for detail.
#' @param pd A data frame of modified sample data file from `MeCall.RemoveBatch()`.
#' @param interest The column names in the sample data file containing the phenotype of interest.
#' @param FDR The false discovery rate (FDR) value applied when identifying DMPs.
#' @param DeltaBeta A vector containing delta beta values for which the user wants to determine the statistical power in a simulation.
#' @param save.plot The directory where the line graph generated by `pwrEWAS` will be saved. If set to `NULL`, the plot will not be created.
#'
#' @return A dataframe containing the statistical power of each delta beta value in the given sample size.
#'
#' @author Hyun-Ho Yang
#'
#' @seealso [pwrEWAS::pwrEWAS()], [MeCall.DMP()]
#'
#' @references
#'
#' Graw, S. et al. (2019). pwrEWAS: a user-friendly tool for comprehensive power estimation for 
#' epigenome wide association studies (EWAS). BMC Bioinformatics. 
#' /url{https://doi.org/10.1186/s12859-019-2804-7}
#'
#' @examples
#' \dontrun{
#' # Simulate statistical power and generate the power plot.
#' data.Power <- MeCall.Power(DMP.object = MeCall.DMP, tissue.type = "Adult (PBMC)", 
#' pd = MeCall.RemoveBatch$modified.pd, interest = "Sample_Group", FDR = 0.05, 
#' DeltaBeta = c(0.02, 0.05, 0.07, 0.1), plot=TRUE)
#' }
#'
#' @export
MeCall.Power <- function(DMP.object = DMPs, tissue.type = "Adult (PBMC)", pd = pd, interest = "Sample_Group",FDR = 0.05, DeltaBeta = NULL, save.plot=NULL){

if(!"pwrEWAS" %in% loadedNamespaces()){
stop("\n[MeCall]-!!ERROR!! : [pwrEWAS] R package is not founded in your R environment. Please load [pwrEWAS] R package to run this function {library(pwrEWAS)}. You can download [pwrEWAS] R package at following github page : https://github.com/stefangraw/pwrEWAS")
}

Available.tissue <- c("Adult (PBMC)", "Saliva", "Sperm", "Lymphoma","Placenta", "Liver", "Colon", "Blood adult", "Blood 5 year olds", "Blood newborns", "Cord-blood (whole blood)", "Cord-blood (PBMC)")

if(!tissue.type %in% Available.tissue){
stop("\n[MeCall]-!!ERROR!! : ", tissue.type, " is not available. Please select different tissue type : 'Adult (PBMC)', 'Saliva', 'Sperm', 'Lymphoma', 'Placenta', 'Liver', 'Colon', 'Blood adult', 'Blood 5 year olds', 'Blood newborns', 'Cord-blood (whole blood)', 'Cord-blood (PBMC)'")
}

if("DMPresultlist" %in% is(DMP.object)){
Dlist <- DMP.object@Main.info
}else{
Need.col <- c("adjusted.p","deltabeta")
if(!all(Need.col %in% colnames(DMP.object))){
missing.one <- Need.col[!Need.col %in% colnames(DMP.object)]
stop("\n[MeCall]-!!ERROR!! : Column name [", paste(missing.one, collapse = ", "),"] are not founded in [DMP.object] parameter. Only exact name matches are allowed. Please check again.")
}else{Dlist <- DMP.object}
}


if(!is.null(save.plot)){

if(str_sub(save.plot, start= -1) != "/"){
save.plot <- paste0(save.plot,"/")
}

if(!dir.exists(save.plot)){
stop("\n[MeCall]-!!ERROR!! : Directory is not founded. Please check again.")
}
}

#### Set detail parameters for calculation of statistical power 
User.Sample.num <- nrow(pd)

min.User.Sample.num <- User.Sample.num-30
max.User.Sample.num <- User.Sample.num+30

if(min.User.Sample.num < 3){
min.User.Sample.num <- 3
}

User.sample.ratio <- table(pd[,interest])[1]/sum(table(pd[,interest]))



Tested.CpG.num <- nrow(Dlist)
Potential.DMPs <- subset(Dlist, adjusted.p <= 0.05)
Potential.DMPs <- nrow(subset(Potential.DMPs, deltabeta < -0.02 | deltabeta > 0.02))

if(is.null(DeltaBeta)){
DeltaBeta <- c(0.02, 0.05, 0.07, 0.1)
}


#### Simulate statistical power given conditions
Statistical_Power <- pwrEWAS(minTotSampleSize = min.User.Sample.num, maxTotSampleSize = max.User.Sample.num,SampleSizeSteps = 10, NcntPer = User.sample.ratio,targetDelta = DeltaBeta, J = Tested.CpG.num, targetDmCpGs = Potential.DMPs, tissueType = tissue.type, detectionLimit = 0.01, DMmethod = "limma", FDRcritVal = FDR, core = 1, sims = 30)

Power.object <-Statistical_Power

#### Generate and save Power plot
if(!is.null(save.plot)){

message("\n[MeCall]-[NOTICE] : MethylCallR will save the power plot at the following directory : ", save.plot)

powerplot <- pwrEWAS_powerPlot(data = Statistical_Power$powerArray, sd = FALSE)
plot.dir <-paste0(save.plot, "EWAS_Power_plot_pwrEWAS.pdf")
grDevices::pdf(plot.dir)
print(powerplot)
grDevices::dev.off()
}

return(Power.object)}
